clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/controllers/GoogleNamedImageLookupController.java

```diff
@@ -0,0 +1,207 @@
+    if (q != null && q.trim().length() > 0) {
+      String glob = q.trim();
+      // Wrap in '*' if there are no glob-style characters in the query string.
+      if (!glob.contains("*") && !glob.contains("?") && !glob.contains("[") && !glob.contains("\\")) {
+        glob = "*" + glob + "*";
+      }
+      Pattern pattern = new InMemoryCache.Glob(glob).toPattern();
+      queryFilter = i-> pattern.matcher(i.imageName).matches();
+    }
+
+    return results.stream()
+      .filter(queryFilter)
+      .filter(namedImage -> matchesTagFilters(namedImage, extractTagFilters(request)))
+      .sorted(Comparator.comparing(image -> image.imageName))
+      .collect(Collectors.toList());
+  }
+
+  private Map<String, List<Image>> listImagesByAccount() {
+    Collection<String> identifiers = cacheView.filterIdentifiers(IMAGES.getNs(), GoogleCloudProvider.getID() + ":*");
+    Map<String, List<Image>> result = new HashMap<>();
+
+    Collection<CacheData> allCacheData = cacheView.getAll(IMAGES.getNs(), identifiers, RelationshipCacheFilter.none());
+    allCacheData.forEach(
+      cacheData -> {
+        String account = Keys.parse(cacheData.getId()).get("account");
+        if (!result.containsKey(account)) {
+          result.put(account, new ArrayList<>());
+        }
+        Object hashImage = cacheData.getAttributes().get("image");
+        try {
+          Image myImage = jsonMapper.fromString(objectMapper.writeValueAsString(hashImage), Image.class);
+          result.get(account).add(myImage);
+        } catch (IOException e) {
+          throw new RuntimeException("Image deserialization failed");
+        }
+      }
+    );
+
+    return result;
+  }
+
+  @VisibleForTesting
+  public static Map<String, String> buildTagsMap(Image image) {
+    Map<String, String> tags = new HashMap<>();
+
+    String description = image.getDescription();
+    if (description != null) {
+      tags = Arrays.stream(description.split(","))
+        .filter(token -> token.contains(": "))
+        .map(token -> token.split(": ", 2))
+        .collect(Collectors.toMap(
+          token -> token[0].trim(),
+          token -> token[1].trim()
+        ));
+    }
+
+    Map<String, String> labels = image.getLabels();
+    if (labels != null) {
+      tags.putAll(labels);
+    }
+
+    return tags;
+  }
+
+  /**
+   * Apply tag-based filtering to the list of named images.
+   *
+   * For example: /gce/images/find?q=PackageName&tag:stage=released&tag:somekey=someval
+   */
+  private static List<NamedImage> filter(List<NamedImage> namedImages, Map<String, String> tagFilters) {
+    return namedImages.stream()
+      .filter(namedImage -> matchesTagFilters(namedImage, tagFilters))
+      .collect(Collectors.toList());
+  }
+
+  private static boolean matchesTagFilters(NamedImage namedImage, Map<String, String> tagFilters) {
+    Map<String, String> tags = namedImage.tags;
+    return tagFilters.keySet().stream()
+      .allMatch(
```

Comment:
Maybe put a `.map(String::toLowerCase)`? Might make the `allMatch` a little easier to read

Reply:
Ah I guess `tagFilters` is case sensitive?

Reply:
Yes, I had exactly that code earlier but broke a unit test, so changed it to what I have now.